Connecting to the testnet

dora-gt opened the pull request:
@wliu @emschwartz 

This is not actually working. Just playing with and figuring out what to do.
`connect-to-xpring.sh` automatically connects our local node to the Xpring test node.

```
./scripts/testnet/connect-to-xpring.sh -f -x -c -s -a -l
```
(too many options lol)

Questions and suggestions:
- `http://test.xpring.tech/api/accounts/xxx` should accept some parameters so that we don't need to add **our** account to the Xpring side. For example, our ILP address and our endpoint URL.
- `http://test.xpring.tech/api/accounts/xxx` should return their ILP address.
- Can nodes handle SSL connections now? I'm not sure.

dora-gt commented:
Ah it seems that localtunnel accepts non-SSL connections though I'm not sure nodes can handle SSLs or not.
https://github.com/localtunnel/localtunnel/issues/105

I'll check the code.

dora-gt commented:
https://github.com/interledger-rs/interledger-rs/blob/master/crates/interledger-http/src/client.rs#L38-L42

dora-gt commented:
memo:
```
05:15 $ curl \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer doradora" \
>     -d '{
>   "username": "doragt2",
>   "ilp_address": "test.dora-gt-2",
>   "asset_code": "XRP",
>   "asset_scale": 9,
>   "min_balance": -1000,
>   "http_incoming_token": "dora-gt-2",
>   "http_outgoing_token": "doragt2:dora-gt-2",
>   "http_endpoint": "https://ioh-test-dora-gt-2.localtunnel.me/ilp",
>   "routing_relation": "peer"
> }' \
> https://ioh-test-dora-gt.localtunnel.me/accounts
curl: (92) HTTP/2 stream 1 was not closed cleanly: REFUSED_STREAM (err 7)
```

dora-gt commented:
Something strange happens.

This works:
```
$ curl https://ioh-test-dora-gt.localtunnel.me/
{"status":"Ready"}
```

This doesn't work:
```
$ curl \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer doradora" \
>     -d '{
>   "username": "doragt2",
>   "ilp_address": "test.dora-gt-2",
>   "asset_code": "XRP",
>   "asset_scale": 9,
>   "min_balance": -1000,
>   "http_incoming_token": "dora-gt-2",
>   "http_outgoing_token": "doragt2:dora-gt-2",
>   "http_endpoint": "https://ioh-test-dora-gt-2.localtunnel.me/ilp",
>   "routing_relation": "peer"
> }' \
> https://ioh-test-dora-gt.localtunnel.me/accounts
curl: (92) HTTP/2 stream 1 was not closed cleanly: REFUSED_STREAM (err 7)
```

It seems that APIs work with HTTP but don't work with HTTPS.

```
curl \
>     -H "Authorization: Bearer doragt:dora-gt" \
>     -H "Content-Type: application/json" \
>     -d "{\"receiver\":\"http://ioh-test-dora-gt-2.localtunnel.me/accounts/doragt2/spsp\",\"source_amount\":500}" \
>     http://ioh-test-dora-gt.localtunnel.me/accounts/doragt/payments
{"delivered_amount":500}
```

gakonst commented on scripts/testnet/account-template.json:
```diff
@@ -0,0 +1,12 @@
+  "http_outgoing_token": "<http_outgoing_token>",
+  "http_endpoint": "<http_endpoint>",
+  "routing_relation": "peer"
+}
```
We also need `settle_to`, `settle_threshold`, `settlement_engine_url` for settlement related functionalities

gakonst commented:
Since our Oct 3rd deadline is approaching, can we get this into a merge-able status? @dora-gt what are your plans?

cc @emschwartz

dora-gt commented:
@gakonst I tried some but it seemed meaningless to work on this with the current version of the testnet node because the version is older than the current `master` branch. I'm waiting for updates on the nodes. I don't have the access to the nodes so I can't update on my own.

dora-gt commented:
I think @wliu is working on it.

dora-gt commented:
![image](https://user-images.githubusercontent.com/811594/65867692-da23b300-e3b1-11e9-95e2-a3be5f2b4152.png)

settlement engine is not done yet.

emschwartz commented on scripts/testnet/connect-to-xpring.sh:
```diff
@@ -0,0 +1,419 @@
+    if [ $? -eq 0 ]; then
+        return 0
+    else
+        error_and_exit "\nNot found. You have to install ${install_name} first."
```
It would be good to provide instructions for how to install it (for example, the command to run to install localtunnel. Alternatively, we could even ask if they want to install it and just run the `npm install -g localtunnel` command for them)

dora-gt replied:
First I tried to make the whole process work really automatically but I finally realized that there are many OSs so I thought that was too hard. Anyway, of course I write a document about how to install these dependencies.

emschwartz commented on scripts/testnet/connect-to-xpring.sh:
```diff
@@ -0,0 +1,419 @@
+
+    # load settings from setting file
+    if [ ! -e "${CONFIG_NAME}" ]; then
+        error_and_exit "Could not find config file: ${CONFIG_NAME}."
```
If I run this for the first time, it logs this error

dora-gt replied:
If we specify the `-c` option, there should be the config file.
If we haven't created the config file, this cannot work because the script doesn't know our admin auth token.
I'll fix to ask the user when the config file is not found.

dora-gt replied:
Did you run like `-c -a`? If so, I don't come up with the cause...

emschwartz replied:
I wasn't sure what commands I was supposed to run to connect to the testnet. "Configure" sounded like it would configure things for me

dora-gt commented:
Btw I think we should provide this function as just CLI command, not shell script.
Shell script is hard to maintain...

dora-gt commented:
Fixed some points.

1. Reduced the number of arguments. Use `-i` to see the commands of payments etc.
    - `-c` Clear config and DB cache.
    - `-s` Spin up settlement engine.
    - `-i` Show instructions (Only available after spinning up the node)
2. Moved account JSONs into `logs/accounts`. See when you want to see raw accounts data.

I think I have to wait the Xpring node to be updated to confirm ETH SE is working or not, I start writing the document.

dora-gt commented:
`-i` works like this:

```
$ ./scripts/testnet/connect-to-xpring.sh -i
You can send payments as follows:

    curl \
        -H "Authorization: Bearer user_ex8p9quh:8etm8bsyeohah" \
        -H "Content-Type: application/json" \
        -d '{"receiver":"RECEIVER SPSP ADDRESS","source_amount":AMOUNT}' \
        http://localhost:7770/accounts/user_ex8p9quh/payments

You can pass the following URL as your SPSP URL:

    https://node-407d2af1eca8a5ca.localtunnel.me/accounts/user_ex8p9quh/spsp

You can also check your balance as follows:

    curl \
        -H "Authorization: Bearer user_ex8p9quh:8etm8bsyeohah" \
        http://localhost:7770/accounts/user_ex8p9quh/balance
```

gakonst commented:
A few notes:
1. I do not think this script should run `cargo compile` at all, just assume that the user has done that. This is pure about connecting to the testnet
2. When I run the script without an argument or with `-h`, it should stop after printing the help prompt. Currently, it continues the initialization process.

dora-gt commented:
`-h` has actually weird behavior... sry. Fixed.

Regarding the compilation, I would like to leave this as-is because if we don't compile first, it could cause compilation sometimes and we don't know how much it takes. Then, `wait_to_serve` would fail. That is not good IMO.

dora-gt commented:
We just need to change this:
https://github.com/interledger-rs/interledger-rs/pull/314/files#diff-774dcab0f8689ea33d2e40bdef48c7a2R489

ryangyoung commented on docs/connect-to-testnet.md:
```diff
@@ -0,0 +1,355 @@
+## Get Your Own Credentials
+You need to get your own credentials to connect to the testnet. In other words, you need to get your account configured on a testnet node.
+
+We have [the Xpring testnet site](https://test.xpring.io/ilp-testnet-creds) to provide credentials for you. The site automatically configures your account on the Xpring's testnet node. You could use the credentials to send payments from the node and receive payments to the node, or you could also connect your local node to the testnet.
```
> [the Xpring testnet site](https://test.xpring.io/ilp-testnet-creds) 

The link should remove `test`, i.e.: https://xpring.io/ilp-testnet-creds

gakonst replied:
@ryangyoung I get a page not found error on this link

ryangyoung replied:
@gakonst Yeah, it's not live quite yet, but it will be as of tomorrow morning PST.

dora-gt replied:
fixed.

dora-gt commented:
Pushed the final version.

Because I changed these lines for tomorrow:
https://github.com/interledger-rs/interledger-rs/pull/314/files#diff-774dcab0f8689ea33d2e40bdef48c7a2R487-R488
this doesn't work now. Care.

gakonst commented:
closing since #370 got merged
